### PR TITLE
gh-112026: Deprecate _PyDict_GetItemStringWithError() function

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-112026

### CODE REVIEW

**Comment by vstinner** [comment]
cc @serhiy-storchaka

File: Include/cpython/dictobject.h [hunk 1]
```diff
@@ -37,7 +37,8 @@ typedef struct {
 
 PyAPI_FUNC(PyObject *) _PyDict_GetItem_KnownHash(PyObject *mp, PyObject *key,
                                                  Py_hash_t hash);
-PyAPI_FUNC(PyObject *) _PyDict_GetItemStringWithError(PyObject *, const char *);
+// PyDict_GetItemStringRef() can be used instead
+Py_DEPRECATED(3.14) PyAPI_FUNC(PyObject *) _PyDict_GetItemStringWithError(PyObject *, const char *);
```
Comment (serhiy-storchaka) [inline Include/cpython/dictobject.h:41]: Maybe 3.13?

**Review by serhiy-storchaka — COMMENTED** [review]
I do not think we should bother with deprecating the internal C API. It can be removed or changed in incompatible way at any time.

File: Include/cpython/dictobject.h (hunk 1 above)
Comment (vstinner) [reply Include/cpython/dictobject.h:41]: I don't like adding new compiler warnings after beta1. There is no urgency to remove it.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
`_PyDict_GetItemStringWithError()` is just a wrapper around `PyDict_GetItemWithError()`. There are no reasons of using it instead of your own wrapper around `PyDict_GetItemWithError()`.

**Comment by vstinner** [comment]
numpy and other projects used it. numpy was updated to use PyDict_GetItemStringRef() using pythoncapi-compat.

**Pull request merged**